### ISSUE TITLE
qualifiers for auto_type in C

### ISSUE THREAD

**Issue by gustedt** [llvm/llvm-project#53652]
Hi,
when preparing for standardization in WG14 of this feature we ran into an inconsistency with gcc.
For declarations with `__auto_type` gcc accepts all C qualifiers in a reasonable way, whereas clang refuses `restrict` and `_Atomic`. The diagnostics that clang gives seem to indicate that these have simply not yet been considered 

> restrict requires a pointer or reference ('__auto_type' is invalid)
> _Atomic cannot be applied to type '__auto_type' which is not trivially copyable

A small example that compiles with gcc but not with clang:

```C
int main() {
  double x = 37;
  // clang doesn't accept that
  __auto_type restrict xr = &x;
  restrict __auto_type rx = &x;
  __auto_type _Atomic xa = x;
   _Atomic __auto_type ax = x;
}
```
Thanks
Jens

**Comment by AaronBallman** [comment]
https://reviews.llvm.org/D12686 was the original review for this feature, and it does not mention leaving these qualifiers out intentionally or test those cases. Given that the patch was intended for GCC compatibility, I agree that this is a bug in Clang.

**Comment by AaronBallman** [comment]
I've put up a patch that hopefully improves things: https://reviews.llvm.org/D122029

**Comment by AaronBallman** [comment]
This should now be addressed by e4a42c5b64d044ae28d9483b0ebd12038d5b5917
